File: app/src/main/java/com/two_two/checkreaction/domain/game/RandomDelayTimer.java
package com.two_two.checkreaction.domain.game;

import android.os.Handler;

import java.util.Random;

/**
 * Created by Dmitry Borodin on 1/4/2016.
 * Manually configured delay timer.
 * Runs provided code after random delay with preconfigured settings.
 */
public final class RandomDelayTimer implements DelayTimer {

    private static final int MIN_DELAY = 500;
    private static final int RANDOM_DELAY1 = 5000;
    private static final int RANDOM_DELAY2 = 1000;
    private final Random mRandom = new Random();

    @Override
    public void runDelayed(final DelayTimerCallback delayTimerCallback) {
        new Handler().postDelayed(delayTimerCallback::delayedCode, getRandomTime());
    }

    private int getRandomTime() {
        //two randoms to more nice delay distribution - less possible min delay
        return mRandom.nextInt(RANDOM_DELAY1) + mRandom.nextInt(RANDOM_DELAY2) + MIN_DELAY;
    }

}

File: app/src/main/java/com/two_two/checkreaction/domain/game/RandomDelayTimer.kt
package com.two_two.checkreaction.domain.game

import android.os.Handler
import android.os.Looper
import java.util.*

/**
 * Created by Dmitry Borodin on 1/4/2016.
 * Manually configured delay timer.
 * Runs provided code after random delay with preconfigured settings.
 */
class RandomDelayTimer : DelayTimer {

    private val random = Random()
    private val handler = Handler(Looper.getMainLooper())
    private lateinit var runnable : Runnable

    override fun runDelayed(delayTimerCallback: DelayTimerCallback) {
        runnable = Runnable{ delayTimerCallback.delayedCode() }
        handler.postDelayed(runnable, randomTime.toLong())
    }

    override fun forgetDelayetCode() {
        handler.removeCallbacks(runnable)
    }

    private //two randoms to more nice delay distribution - less possible min delay
    val randomTime: Int
        get() = random.nextInt(RANDOM_DELAY1) + random.nextInt(RANDOM_DELAY2) + MIN_DELAY

    companion object {
        private const val MIN_DELAY = 500
        private const val RANDOM_DELAY1 = 5000
        private const val RANDOM_DELAY2 = 1000
    }

}
